src/porepy/numerics/time_step_control.py

```diff
@@ -0,0 +1,273 @@
+
+        # Time step. Initially, equal to the initial time step
+        self.dt = self.dt_init
+
+        # Private attributes
+        # Number of times the solution has been recomputed
+        self._recomp_num = 0
+
+        # Index of the next scheduled time
+        self._scheduled_idx = 1
+
+        # Print information
+        self._print_info = print_info
+
+    def __repr__(self) -> str:
+
+        s = "Time-stepping control object with atributes:\n"
+        s += f"Initial simulation time = {self.time_init}\n"
+        s += f"Final simulation time = {self.time_final}\n"
+        s += f"Initial time step = {self.dt_init}\n"
+        s += f"Minimum time step = {self.dt_min}\n"
+        s += f"Maximum time step = {self.dt_max}\n"
+        s += f"Lower optimal iteration range = {self.iter_low}\n"
+        s += f"Upper optimal iteration range = {self.iter_upp}\n"
+        s += f"Lower multiplication factor = {self.iter_low_factor}\n"
+        s += f"Upper multiplication factor = {self.iter_upp_factor}\n"
+        s += f"Recompute solution multiplication factor = {self.recomp_factor}\n"
+        s += f"Maximum recomputing attempts = {self.recomp_max}"
+
+        return s
+
+    def next_time_step(
+        self, iterations: int, recompute_solution: bool
+    ) -> Union[float, None]:
+        """
+        Determines the next time step based on the previous amount of iterations needed
+        to reach convergence. If convergence was not achieved, then the time step is
+        reduced by recomp_factor. The time-stepping control routine will recompute the
+        solution recomp_max times. Otherwise, an error will be raised and the simulation
+        stopped.
+
+        Parameters
+        iterations (int): Number of non-linear iterations. In time-dependent simulations,
+            this tipically represent the number of iterations for a given time step.
+        recompute_solution (bool): Wheter the solution needs to be recomputed or not. If
+            True, then the time step is multiplied by recomp_factor. If False, the time
+            step will be tune accordingly.
+
+        Returns
+        dt (float or None):  Next time step if time < final_simulation time. None otherwise.
+
+        """
+
+        # First, check if we reach final simulation time
+        if self.time == self.time_final:
+            return None
+
+        # If the solution did not convergence and we are allow to recompute it:
+        #   Update simulation time (since solution will be recomputed)
+        #   Decrease time step by the recomputing factor
+        #   Increase counter keeping track of the number of times the solution was recomputed
+        #   Check if the time step is smaller that dt_min. Otherwise, use dt_min
+        if recompute_solution and self._recomp_num < self.recomp_max:
+            self.time -= self.dt
+            self.dt *= self.recomp_factor
+            self._recomp_num += 1
+            if self._print_info:
```

Comment:
I'd prefer to have this by logging, but that could also wait till a later revision of the code.

Reply:
I agree, not very elegant the way is currently handled. But yes, let's postpone this fix for the (hopefully near) future.